[DataLoader] Typing Enforcement for DataPipe at Compile-time

ejguan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54773 [DataLoader] Typing Doc
* #54544 [DataLoader] Typing Enforcement for DataPipe at runtime
* #54066 [DataLoader] Typing Enforcement for DataPipe at construct-time
* **#54020 [DataLoader] Typing Enforcement for DataPipe at Compile-time**
* #54299 [DataLoader] Implement issubtype for type hints


## Feature
- Add `issubtype` to check the type is a subtype of the other type.
- Add `_DataPipeMeta` (mimic Python typing 3.6)
  - Add `type` attribute for each DataPipe
  - Validate return hint of `__iter__`

No Error for main repo, `torchvision`, `torchaudio` and `torchtext`.

## Next Step
- When DataFrame came out, add an another type for DataFrame with column name and type.
- When Python 3.6 is deprecated, this part can be optimized by TypeAlias by reducing inheritance path and reduce costs for `__new__` for each Meta. Also, it will make DataPipe typing system more consistent with Python typing system.

## BC Breaking Note
This PR introduces a new attributed named `type` for `IterableDataset` using the typing annotation attached to each `IterableDataset`. By having this attribute, we are able to extend `IterableDataset` to have type inference and lazy initialization to incorporate the new DataLoader architecture. So, it introduces several BC-breaking cases as shown below.

### 1.8.1
The following code can be executed without any error with PyTorch 1.8.1
```py
class DP(IterableDataset["invalid_type"]):
    def __iter__(self) -> "invalid_type":
        ...

class DP(IterableDataset[str]):
    def __iter__(self) -> str:
        ...

class DP(IterableDataset[str]):
    def __iter__(self) -> Iterator[int]:
        ...
```

### This PR
1. Raise Exception where `IterableDataset` is annotated with a string of invalid type.
The type annotation is required to be valid now. Previous workaround like using string to represent the invalid type annotation is not supported after this PR.

2. Raise `TypeError` if the return type of `__iter__` is not an `Iterator `

3. Raise `TypeError` if the return type of `__iter__` is of the form `Iterator[X]`,  but the argument type `X` is not a subtype of the `IterableDataset.type` attribute.

4. `IterableDatset` now has a metaclass, which can conflict with existing user-defined metaclasses on `IterableDatasets`.

```py
# 1
class DS(IterableDataset["invalid_type"]):  # Raise Exception from the evaluation `eval("invalid_type", globals, locals)`
    def __iter__(self) -> "invalid_type":
        ...

# 2
class DS(IterableDataset[str]):
    def __iter__(self) -> str:  # Raise `TypeError: Expected 'Iterator' as the return annotation for '__iter__' of DS, but found str`
        ...

# 3
class DS(IterableDataset[str]):
    def __iter__(self) -> Iterator[int]:  # Raise `TypeError: Expected return type of '__iter__' is a subtype of str, but found int`
        ...

# 4
class DS(IterableDataset[str], metaclass=MyMeta):  # Raise `TypeError: metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass of the metaclasses of all its bases`
    pass
```

Differential Revision: [D27327232](https://our.internmc.facebook.com/intern/diff/D27327232)

ejguan committed:
[DataLoader] Typing Enforcement for DataPipe at Compile-time

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at Compile-time"




[ghstack-poisoned]

ejguan commented on torch/utils/data/dataset.py:
```diff
@@ -174,6 +176,21 @@ def class_function(cls, source_dp, *args, **kwargs):
         function = functools.partial(class_function, cls_to_register)
         IterableDataset.functions[function_name] = function
 
+    # TODO: Determinin if force all IterableDataset and IterDataPipe with annotation
+    #  def __init_subclass__(cls, *args, **kwargs):
+    #      if not hasattr(cls, 'type'):
+    #          raise TypeError('Class {} needs to be specified with type'.format(cls.__name__))
```
In case of BC-breaking, IMO it should be commented out. There are lots of Dataset not specifying Type.

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at Compile-time"


No Error for main repo, `torchvision`, `torchaudio` and `torchtext`.

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at Compile-time"


No Error for main repo, `torchvision`, `torchaudio` and `torchtext`.

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at Compile-time"


No Error for main repo, `torchvision`, `torchaudio` and `torchtext`.
Typing for DataPipe is hard to handle as it includes `TypeVar` and `...`, which introduce not fixed type.

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at Compile-time"


No Error for main repo, `torchvision`, `torchaudio` and `torchtext`.
Typing for DataPipe is hard to handle as it includes `TypeVar` and `...`, which introduce not fixed type.

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at Compile-time"


No Error for main repo, `torchvision`, `torchaudio` and `torchtext`.
Typing for DataPipe is hard to handle as it includes `TypeVar` and `...`, which introduce not fixed type.

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at Compile-time"


No Error for main repo, `torchvision`, `torchaudio` and `torchtext`.
Typing for DataPipe is hard to handle as it includes `TypeVar` and `...`, which introduce not fixed type.

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at Compile-time"


No Error for main repo, `torchvision`, `torchaudio` and `torchtext`.
Typing for DataPipe is hard to handle as it includes `TypeVar` and `...`, which introduce not fixed type.

[ghstack-poisoned]

ejguan commented on torch/utils/data/dataset.py:
```diff
@@ -189,6 +191,21 @@ def set_reduce_ex_hook(cls, hook_fn):
             raise Exception("Attempt to override existing reduce_ex_hook")
         IterableDataset.reduce_ex_hook = hook_fn
 
+    #  def __init_subclass__(cls, *args, **kwargs):
+    #      # TODO: Determinin if force all IterableDataset and IterDataPipe with annotation
+    #      if not hasattr(cls, 'type'):
+    #          raise TypeError('Class {} needs to be specified with type'.format(cls.__name__))
+
+    @_tp_cache
+    def __class_getitem__(cls, param) -> _DataPipeAlias:
+        # TODO: add global switch for type checking at compile-time
+        if param is None:
+            raise TypeError('Can not take None as type parameter for {}'.format(cls.__name__))
+        if isinstance(param, Sequence):
+            param = Tuple[param]
+        _type_check(param, msg="Parameters for {} must be types".format(cls.__name__))
+        return _DataPipeAlias(cls, param)
+
```
This part can be moved to `DataSet` to make all previous map-stype Dataset gains the benefits (?) from more strict typing requirement.

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at Compile-time"


## Workflow
- Create `DataPipe` with subscribed type(e.g. `class ABCDataPipe(IterDataPipe[int])`)
- If the type has been met before, use cache.
- If the type has never been met, create a new `_DataPipeAlias` 
  - Create type and determine if the type is a fixed or non-fixed type
    - all cases need to check return type of `__iter__` is same as class annotation (e.g. int as above)
    - non-fixed type to create new `__init__` function to copy the type variable for each instance.

No Error for main repo, `torchvision`, `torchaudio` and `torchtext`.

## TODO
When DataFrame came out, add a another type for DataFrame with column name and type.

[ghstack-poisoned]

ejguan commented on torch/utils/data/typing.py:
```diff
@@ -147,3 +158,179 @@ def _issubtype_with_constraints(variant, constraints):
+                raise TypeError('Expected Iterator as the return annotation for `__iter__` of {}'
+                                ', but found {}'.format(sub_cls.__name__, hints['return']))
+            data_type = return_hint.__args__[0]
+            if not (issubtype(sub_cls.type.param, data_type) and issubtype(data_type, sub_cls.type.param)):
```
Double side subtype can find equal typing. e.g. `Any == T_co`

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at Compile-time"


## Workflow
- Create `DataPipe` with subscribed type(e.g. `class ABCDataPipe(IterDataPipe[int])`)
- If the type has been met before, use cache.
- If the type has never been met, create a new `_DataPipeAlias` 
  - Create type and determine if the type is a fixed or non-fixed type
    - all cases need to check return type of `__iter__` is same as class annotation (e.g. int as above)
    - non-fixed type to create new `__init__` function to copy the type variable for each instance.

No Error for main repo, `torchvision`, `torchaudio` and `torchtext`.

## TODO
When DataFrame came out, add a another type for DataFrame with column name and type.

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at Compile-time"


## Workflow
- Create `DataPipe` with subscribed type(e.g. `class ABCDataPipe(IterDataPipe[int])`)
- If the type has been met before, use cache.
- If the type has never been met, create a new `_DataPipeAlias` 
  - Create type and determine if the type is a fixed or non-fixed type
    - all cases need to check return type of `__iter__` is same as class annotation (e.g. int as above)
    - non-fixed type to create new `__init__` function to copy the type variable for each instance.

No Error for main repo, `torchvision`, `torchaudio` and `torchtext`.

## TODO
When DataFrame came out, add a another type for DataFrame with column name and type.

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at Compile-time"


## Workflow
- Create `DataPipe` with subscribed type(e.g. `class ABCDataPipe(IterDataPipe[int])`)
- If the type has been met before, use cache.
- If the type has never been met, create a new `_DataPipeAlias` 
  - Create type and determine if the type is a fixed or non-fixed type
    - all cases need to check return type of `__iter__` is same as class annotation (e.g. int as above)
    - non-fixed type to create new `__init__` function to copy the type variable for each instance.

No Error for main repo, `torchvision`, `torchaudio` and `torchtext`.

## TODO
When DataFrame came out, add a another type for DataFrame with column name and type.

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at Compile-time"


## Workflow
- Create `DataPipe` with subscribed type(e.g. `class ABCDataPipe(IterDataPipe[int])`)
- If the type has been met before, use cache.
- If the type has never been met, create a new `_DataPipeAlias` 
  - Create type and determine if the type is a fixed or non-fixed type
    - all cases need to check return type of `__iter__` is same as class annotation (e.g. int as above)
    - non-fixed type to create new `__init__` function to copy the type variable for each instance.

No Error for main repo, `torchvision`, `torchaudio` and `torchtext`.

## TODO
When DataFrame came out, add a another type for DataFrame with column name and type.

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at Compile-time"


## Workflow
- Create `DataPipe` with subscribed type(e.g. `class ABCDataPipe(IterDataPipe[int])`)
- If the type has been met before, use cache.
- If the type has never been met, create a new `_DataPipeAlias` 
  - Create type and determine if the type is a fixed or non-fixed type
    - all cases need to check return type of `__iter__` is same as class annotation (e.g. int as above)
    - non-fixed type to create new `__init__` function to copy the type variable for each instance.

No Error for main repo, `torchvision`, `torchaudio` and `torchtext`.

## TODO
When DataFrame came out, add a another type for DataFrame with column name and type.

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at Compile-time"



## Feature
- Add `issubtype` to check the type is a subtype of the other type.
- Add `_DataPipeMeta` (mimic Python typing 3.6)
  - Add `type` attribute for each DataPipe
  - Save original `__init__` function for each DataPipe
  - Validate return hint of `__iter__`
  - Replace `__init__` function bases on `type`
    - Fixed type: Put original `__init__` back, if it exists or use a plain `__init__`
    -  Non-fixed type: Add new `__init__` with the functionality to copy `cls.type` for each instance. (Optimized for memory)

No Error for main repo, `torchvision`, `torchaudio` and `torchtext`.

## Future
- Add same thing for `__getitem__`.
- When DataFrame came out, add an another type for DataFrame with column name and type.

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at Compile-time"



## Feature
- Add `issubtype` to check the type is a subtype of the other type.
- Add `_DataPipeMeta` (mimic Python typing 3.6)
  - Add `type` attribute for each DataPipe
  - Save original `__init__` function for each DataPipe
  - Validate return hint of `__iter__`
  - Replace `__init__` function bases on `type`
    - Fixed type: Put original `__init__` back, if it exists or use a plain `__init__`
    -  Non-fixed type: Add new `__init__` with the functionality to copy `cls.type` for each instance. (Optimized for memory)

No Error for main repo, `torchvision`, `torchaudio` and `torchtext`.

## Next Step
- Add a flag to switch on/off.
- Add same thing for `__getitem__`.
- When DataFrame came out, add an another type for DataFrame with column name and type.
- When Python 3.6 is deprecated, this part can be optimized by TypeAlias by reducing inheritance path and reduce costs for `__new__` for each Meta.

Differential Revision: [D27327232](https://our.internmc.facebook.com/intern/diff/D27327232)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at Compile-time"



## Feature
- Add `issubtype` to check the type is a subtype of the other type.
- Add `_DataPipeMeta` (mimic Python typing 3.6)
  - Add `type` attribute for each DataPipe
  - Save original `__init__` function for each DataPipe
  - Validate return hint of `__iter__`
  - Replace `__init__` function bases on `type`
    - Fixed type: Put original `__init__` back, if it exists or use a plain `__init__`
    -  Non-fixed type: Add new `__init__` with the functionality to copy `cls.type` for each instance. (Optimized for memory)

No Error for main repo, `torchvision`, `torchaudio` and `torchtext`.

## Next Step
- Add a flag to switch on/off.
- Add same thing for `__getitem__`.
- When DataFrame came out, add an another type for DataFrame with column name and type.
- When Python 3.6 is deprecated, this part can be optimized by TypeAlias by reducing inheritance path and reduce costs for `__new__` for each Meta. Also, it will make DataPipe typing system more consistent with Python typing system.

Differential Revision: [D27327232](https://our.internmc.facebook.com/intern/diff/D27327232)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at Compile-time"



## Feature
- Add `issubtype` to check the type is a subtype of the other type.
- Add `_DataPipeMeta` (mimic Python typing 3.6)
  - Add `type` attribute for each DataPipe
  - Save original `__init__` function for each DataPipe
  - Validate return hint of `__iter__`
  - Replace `__init__` function bases on `type`
    - Fixed type: Put original `__init__` back, if it exists or use a plain `__init__`
    -  Non-fixed type: Add new `__init__` with the functionality to copy `cls.type` for each instance. (Optimized for memory)

No Error for main repo, `torchvision`, `torchaudio` and `torchtext`.

## Next Step
- Add a flag to switch on/off.
- Add same thing for `__getitem__`.
- When DataFrame came out, add an another type for DataFrame with column name and type.
- When Python 3.6 is deprecated, this part can be optimized by TypeAlias by reducing inheritance path and reduce costs for `__new__` for each Meta. Also, it will make DataPipe typing system more consistent with Python typing system.

Differential Revision: [D27327232](https://our.internmc.facebook.com/intern/diff/D27327232)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at Compile-time"



## Feature
- Add `issubtype` to check the type is a subtype of the other type.
- Add `_DataPipeMeta` (mimic Python typing 3.6)
  - Add `type` attribute for each DataPipe
  - Save original `__init__` function for each DataPipe
  - Validate return hint of `__iter__`
  - Replace `__init__` function bases on `type`
    - Fixed type: Put original `__init__` back, if it exists or use a plain `__init__`
    -  Non-fixed type: Add new `__init__` with the functionality to copy `cls.type` for each instance. (Optimized for memory)

No Error for main repo, `torchvision`, `torchaudio` and `torchtext`.

## Next Step
- Add a flag to switch on/off.
- Add same thing for `__getitem__`.
- When DataFrame came out, add an another type for DataFrame with column name and type.
- When Python 3.6 is deprecated, this part can be optimized by TypeAlias by reducing inheritance path and reduce costs for `__new__` for each Meta. Also, it will make DataPipe typing system more consistent with Python typing system.

Differential Revision: [D27327232](https://our.internmc.facebook.com/intern/diff/D27327232)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at Compile-time"



## Feature
- Add `issubtype` to check the type is a subtype of the other type.
- Add `_DataPipeMeta` (mimic Python typing 3.6)
  - Add `type` attribute for each DataPipe
  - Save original `__init__` function for each DataPipe
  - Validate return hint of `__iter__`
  - Replace `__init__` function bases on `type`
    - Fixed type: Put original `__init__` back, if it exists or use a plain `__init__`
    -  Non-fixed type: Add new `__init__` with the functionality to copy `cls.type` for each instance. (Optimized for memory)

No Error for main repo, `torchvision`, `torchaudio` and `torchtext`.

## Next Step
- Add a flag to switch on/off.
- Add same thing for `__getitem__`.
- When DataFrame came out, add an another type for DataFrame with column name and type.
- When Python 3.6 is deprecated, this part can be optimized by TypeAlias by reducing inheritance path and reduce costs for `__new__` for each Meta. Also, it will make DataPipe typing system more consistent with Python typing system.

Differential Revision: [D27327232](https://our.internmc.facebook.com/intern/diff/D27327232)

[ghstack-poisoned]

ejguan commented:
Even if OSS is working, the problem is the internal usage. We have `AutoOnBoxDataset ` using `metaclass`, which raises metaclass conflict. We may need to remove Python 3.6 from this typing system, and use TypeAlias to get rid of metaclass conflicts.

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at Compile-time"



## Feature
- Add `issubtype` to check the type is a subtype of the other type.
- Add `_DataPipeMeta` (mimic Python typing 3.6)
  - Add `type` attribute for each DataPipe
  - Save original `__init__` function for each DataPipe
  - Validate return hint of `__iter__`
  - Replace `__init__` function bases on `type`
    - Fixed type: Put original `__init__` back, if it exists or use a plain `__init__`
    -  Non-fixed type: Add new `__init__` with the functionality to copy `cls.type` for each instance. (Optimized for memory)

No Error for main repo, `torchvision`, `torchaudio` and `torchtext`.

## Next Step
- Add a flag to switch on/off.
- Add same thing for `__getitem__`.
- When DataFrame came out, add an another type for DataFrame with column name and type.
- When Python 3.6 is deprecated, this part can be optimized by TypeAlias by reducing inheritance path and reduce costs for `__new__` for each Meta. Also, it will make DataPipe typing system more consistent with Python typing system.

Differential Revision: [D27327232](https://our.internmc.facebook.com/intern/diff/D27327232)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at Compile-time"



## Feature
- Add `issubtype` to check the type is a subtype of the other type.
- Add `_DataPipeMeta` (mimic Python typing 3.6)
  - Add `type` attribute for each DataPipe
  - Save original `__init__` function for each DataPipe
  - Validate return hint of `__iter__`
  - Replace `__init__` function bases on `type`
    - Fixed type: Put original `__init__` back, if it exists or use a plain `__init__`
    -  Non-fixed type: Add new `__init__` with the functionality to copy `cls.type` for each instance. (Optimized for memory)

No Error for main repo, `torchvision`, `torchaudio` and `torchtext`.

## Next Step
- Add a flag to switch on/off.
- Add same thing for `__getitem__`.
- When DataFrame came out, add an another type for DataFrame with column name and type.
- When Python 3.6 is deprecated, this part can be optimized by TypeAlias by reducing inheritance path and reduce costs for `__new__` for each Meta. Also, it will make DataPipe typing system more consistent with Python typing system.

Differential Revision: [D27327232](https://our.internmc.facebook.com/intern/diff/D27327232)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at Compile-time"



## Feature
- Add `issubtype` to check the type is a subtype of the other type.
- Add `_DataPipeMeta` (mimic Python typing 3.6)
  - Add `type` attribute for each DataPipe
  - Validate return hint of `__iter__`

No Error for main repo, `torchvision`, `torchaudio` and `torchtext`.

## Next Step
- Add a flag to switch on/off.
- Add same thing for `__getitem__`.
- When DataFrame came out, add an another type for DataFrame with column name and type.
- When Python 3.6 is deprecated, this part can be optimized by TypeAlias by reducing inheritance path and reduce costs for `__new__` for each Meta. Also, it will make DataPipe typing system more consistent with Python typing system.

Differential Revision: [D27327232](https://our.internmc.facebook.com/intern/diff/D27327232)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at Compile-time"



## Feature
- Add `issubtype` to check the type is a subtype of the other type.
- Add `_DataPipeMeta` (mimic Python typing 3.6)
  - Add `type` attribute for each DataPipe
  - Validate return hint of `__iter__`

No Error for main repo, `torchvision`, `torchaudio` and `torchtext`.

## Next Step
- Add a flag to switch on/off.
- Make `type` as a read-only property
- Add same thing for `__getitem__`.
- When DataFrame came out, add an another type for DataFrame with column name and type.
- When Python 3.6 is deprecated, this part can be optimized by TypeAlias by reducing inheritance path and reduce costs for `__new__` for each Meta. Also, it will make DataPipe typing system more consistent with Python typing system.

Differential Revision: [D27327232](https://our.internmc.facebook.com/intern/diff/D27327232)

[ghstack-poisoned]

ejguan commented on torch/utils/data/_typing.py:
```diff
@@ -141,3 +154,143 @@ def _issubtype_with_constraints(variant, constraints):
+    # - add global switch for type checking at compile-time
+    if '__iter__' in sub_cls.__dict__:
+        iter_fn = sub_cls.__dict__['__iter__']
+        try:
+            hints = get_type_hints(iter_fn)
+        # Ignore TypeError causes by invalid string typing annotation in order to keep internal BC
+        except TypeError:
+            return
```
There are bugs of typing in internal codebase, have to ignore the exception for non-BC breaking

VitalyFedyunin replied:
Please file issue for it and add warning.

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at Compile-time"



## Feature
- Add `issubtype` to check the type is a subtype of the other type.
- Add `_DataPipeMeta` (mimic Python typing 3.6)
  - Add `type` attribute for each DataPipe
  - Validate return hint of `__iter__`

No Error for main repo, `torchvision`, `torchaudio` and `torchtext`.

## Next Step
- Add a flag to switch on/off.
- Make `type` as a read-only property
- Add same thing for `__getitem__`.
- When DataFrame came out, add an another type for DataFrame with column name and type.
- When Python 3.6 is deprecated, this part can be optimized by TypeAlias by reducing inheritance path and reduce costs for `__new__` for each Meta. Also, it will make DataPipe typing system more consistent with Python typing system.

Differential Revision: [D27327232](https://our.internmc.facebook.com/intern/diff/D27327232)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at Compile-time"



## Feature
- Add `issubtype` to check the type is a subtype of the other type.
- Add `_DataPipeMeta` (mimic Python typing 3.6)
  - Add `type` attribute for each DataPipe
  - Validate return hint of `__iter__`

No Error for main repo, `torchvision`, `torchaudio` and `torchtext`.

## Next Step
- Add a flag to switch on/off.
- Make `type` as a read-only property
- Add same thing for `__getitem__`.
- When DataFrame came out, add an another type for DataFrame with column name and type.
- When Python 3.6 is deprecated, this part can be optimized by TypeAlias by reducing inheritance path and reduce costs for `__new__` for each Meta. Also, it will make DataPipe typing system more consistent with Python typing system.

Differential Revision: [D27327232](https://our.internmc.facebook.com/intern/diff/D27327232)

[ghstack-poisoned]

VitalyFedyunin commented on torch/utils/data/_typing.py:
```diff
@@ -1,5 +1,21 @@
+try:
+    from typing import GenericMeta  # Python 3.6
+except ImportError:  # Python > 3.6
+    # In case of metaclass conflict due to ABCMeta or _ProtocolMeta
+    # For Python 3.9, only Protocol in typing uses metaclass
+    from abc import ABCMeta
+    from typing import _ProtocolMeta  # type: ignore
+
+    class GenericMeta(_ProtocolMeta, ABCMeta):  # type: ignore
+        pass
```
Please add todo

VitalyFedyunin reviewed (approved):
Looks good. I wonder how much readable is `datapipe.type` field and can we make it part of our API (for debugability).

ejguan commented:
> Looks good. I wonder how much readable is `datapipe.type` field and can we make it part of our API (for debugability).

In current setting, all datapipe own a type. Default type will be typing.Any

So I think we can make it in our API. It's useful for some datapipe with strict input requirements. And do lazy_init based the specified type.

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at Compile-time"



## Feature
- Add `issubtype` to check the type is a subtype of the other type.
- Add `_DataPipeMeta` (mimic Python typing 3.6)
  - Add `type` attribute for each DataPipe
  - Validate return hint of `__iter__`

No Error for main repo, `torchvision`, `torchaudio` and `torchtext`.

## Next Step
- Add a flag to switch on/off.
- Make `type` as a read-only property
- Add same thing for `__getitem__`.
- When DataFrame came out, add an another type for DataFrame with column name and type.
- When Python 3.6 is deprecated, this part can be optimized by TypeAlias by reducing inheritance path and reduce costs for `__new__` for each Meta. Also, it will make DataPipe typing system more consistent with Python typing system.

Differential Revision: [D27327232](https://our.internmc.facebook.com/intern/diff/D27327232)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at Compile-time"



## Feature
- Add `issubtype` to check the type is a subtype of the other type.
- Add `_DataPipeMeta` (mimic Python typing 3.6)
  - Add `type` attribute for each DataPipe
  - Validate return hint of `__iter__`

No Error for main repo, `torchvision`, `torchaudio` and `torchtext`.

## Next Step
- Add a flag to switch on/off.
- Make `type` as a read-only property
- Add same thing for `__getitem__`.
- When DataFrame came out, add an another type for DataFrame with column name and type.
- When Python 3.6 is deprecated, this part can be optimized by TypeAlias by reducing inheritance path and reduce costs for `__new__` for each Meta. Also, it will make DataPipe typing system more consistent with Python typing system.

Differential Revision: [D27327232](https://our.internmc.facebook.com/intern/diff/D27327232)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at Compile-time"



## Feature
- Add `issubtype` to check the type is a subtype of the other type.
- Add `_DataPipeMeta` (mimic Python typing 3.6)
  - Add `type` attribute for each DataPipe
  - Validate return hint of `__iter__`

No Error for main repo, `torchvision`, `torchaudio` and `torchtext`.

## Next Step
- Add a flag to switch on/off.
- Make `type` as a read-only property
- Add same thing for `__getitem__`.
- When DataFrame came out, add an another type for DataFrame with column name and type.
- When Python 3.6 is deprecated, this part can be optimized by TypeAlias by reducing inheritance path and reduce costs for `__new__` for each Meta. Also, it will make DataPipe typing system more consistent with Python typing system.

Differential Revision: [D27327232](https://our.internmc.facebook.com/intern/diff/D27327232)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at Compile-time"



## Feature
- Add `issubtype` to check the type is a subtype of the other type.
- Add `_DataPipeMeta` (mimic Python typing 3.6)
  - Add `type` attribute for each DataPipe
  - Validate return hint of `__iter__`

No Error for main repo, `torchvision`, `torchaudio` and `torchtext`.

## Next Step
- Add a flag to switch on/off.
- Make `type` as a read-only property
- Add same thing for `__getitem__`.
- When DataFrame came out, add an another type for DataFrame with column name and type.
- When Python 3.6 is deprecated, this part can be optimized by TypeAlias by reducing inheritance path and reduce costs for `__new__` for each Meta. Also, it will make DataPipe typing system more consistent with Python typing system.

Differential Revision: [D27327232](https://our.internmc.facebook.com/intern/diff/D27327232)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Enforcement for DataPipe at Compile-time"



## Feature
- Add `issubtype` to check the type is a subtype of the other type.
- Add `_DataPipeMeta` (mimic Python typing 3.6)
  - Add `type` attribute for each DataPipe
  - Validate return hint of `__iter__`

No Error for main repo, `torchvision`, `torchaudio` and `torchtext`.

## Next Step
- Add a flag to switch on/off.
- Make `type` as a read-only property
- Add same thing for `__getitem__`.
- When DataFrame came out, add an another type for DataFrame with column name and type.
- When Python 3.6 is deprecated, this part can be optimized by TypeAlias by reducing inheritance path and reduce costs for `__new__` for each Meta. Also, it will make DataPipe typing system more consistent with Python typing system.

Differential Revision: [D27327232](https://our.internmc.facebook.com/intern/diff/D27327232)

[ghstack-poisoned]

ezyang commented:
Some after the fact commentary.

First, I personally find the BC note very hard to understand.  First, some blow-by-blow grammar problems:

* "The following generic namedtuple is supported after Python3.7." - why do I care about the following generic namedtuple? Is it an example, is it part of the API? What's going on with Python 3.7? Is it supported in Python 3.7, or only in 3.8? Why is it only supported later in Python?
* "But, since it drops the Generic internally, user can not subscript this class like GenericNamedTuple[torch.Tensor]." What does "it" refer to here? Why do I care about it dropping the Generic? What does it mean to "drop" a Generic? What is the "but" referring to? Why do I care about being able to subscript the class; this class is nowhere in this PR, why is this at all relevant to PyTorch?
* "But users can still use a string to represent it like "GenericNamedTuple[torch.Tensor]", which will create a ForwardRef as type annotation without any error." You used "but" a second time. Is this negating the but, or a continuation of the previous statement? OK, they can use a string. Why do I care about this creating a "ForwardRef"? What's a "ForwardRef"? Is this trying to explain a workaround?

Part of what seems to be going on is that you are trying to explain the BC implications of a metaclass addition. This is generally fairly difficult, since this isn't a traditional Python API boundary and adding metaclasses where there weren't any before always leads to weird knock on effects. But then just say **we added a metaclass, some common things that may stop working are X, Y and Z**.

Second, I don't really understand why this had be rushed in prior to dropping support for Python 3.6 (which we should be clear to do now, we support three most recent versions of Python 3.7, 3.8 and 3.9). It seems to me like the implementation here is substantially simpler, and less BC-breaking, when you don't actually need to deal with GenericMeta. No one is going to die if their code is slightly less typechecked for a few more months. What actually is the motivation behind all of this machinery?

ejguan commented:
@ezyang 
Just want to make sure you have more context of the reason for this typing system:
1. We need the type attribute for some `DataPipe`s to achieve loading data lazily. For example, `UnZipDataPipe` needs to know the output tuple size of the prior input `DataPipe`. With the type information of the prior DataPipe, we are able to split one `DataPipe` into a corresponding number of `DataPipe`s. Otherwise, we have to load the first piece of data from the prior `DataPipe` to achieve this function.
2. This type system is a way to achieve the functionality like data extraction and validation for the DataFrame-like `DataPipe` (torcharrow).
```py
class DP:
    def __init__(self, dp: DataFrameDataPipe["name": str, "val": float]):
        ...
```
3. Static/Runtime type check

> - "The following generic namedtuple is supported after Python3.7." - why do I care about the following generic namedtuple? Is it an example, is it part of the API? What's going on with Python 3.7? Is it supported in Python 3.7, or only in 3.8? Why is it only supported later in Python?
> - "But, since it drops the Generic internally, user can not subscript this class like GenericNamedTuple[torch.Tensor]." What does "it" refer to here? Why do I care about it dropping the Generic? What does it mean to "drop" a Generic? What is the "but" referring to? Why do I care about being able to subscript the class; this class is nowhere in this PR, why is this at all relevant to PyTorch?
> - "But users can still use a string to represent it like "GenericNamedTuple[torch.Tensor]", which will create a ForwardRef as type annotation without any error." You used "but" a second time. Is this negating the but, or a continuation of the previous statement? OK, they can use a string. Why do I care about this creating a "ForwardRef"? What's a "ForwardRef"? Is this trying to explain a workaround?

You are right, I made the Note too confusing as I explained too much detailed information.

In general, I want to explain **all** invalid type annotation is not supported now. Before this PR, users can use a workaround like convert the invalid type annotation as a string to ignore it.

And, I use the generic namedtuple as a special example, which has different behavior across different Python version with same code.
- 1.8.1
1. Python 3.6
Error out when we define `GenericNamedTuple`
```py
class GenericNamedTuple(Generic[T_co], NamedTuple):  # TypeError: metaclass conflict
    name: str
    data: T_co
```
2. Python >= 3.6
Python allows the generic namedtuple. Even though Python doesn't allow subscript it directly like `GenericNamedTuple[int]`, user can make it into a string (`ForwardRef`) as the annotation to work around this problem like the following code without any Error.
```py
class GenericNamedTuple(Generic[T_co], NamedTuple):
    name: str
    data: T_co
class DP(IterableDataset["GenericNamedTuple[torch.Tensor]"]):  # Use string
   ...
```

- This PR
1. Python 3.6
Still error out when we define `GenericNamedTuple`. There won't be any BC-breaking in this case.
```py
class GenericNamedTuple(Generic[T_co], NamedTuple):  # TypeError: metaclass conflict
    name: str
    data: T_co
```
2. Python >= 3.6
Since all `IterDataPipe` (`IterableDataset`) now require valid type with this PR, the following code will error out even if we use string to represent the invalid type.
```py
class GenericNamedTuple(Generic[T_co], NamedTuple):
    name: str
    data: T_co
class DP(IterableDataset["GenericNamedTuple[torch.Tensor]"]):  # TypeError: 'type' object is not subscriptable
   ...
```

I am going to update the BC-breaking Note to make it clear.

> Second, I don't really understand why this had be rushed in prior to dropping support for Python 3.6 (which we should be clear to do now, we support three most recent versions of Python 3.7, 3.8 and 3.9). It seems to me like the implementation here is substantially simpler, and less BC-breaking, when you don't actually need to deal with GenericMeta. No one is going to die if their code is slightly less typechecked for a few more months. What actually is the motivation behind all of this machinery?

When I investigate and design this typing system, I have the same feeling for dropping support for Python 3.6. But, based on the schedule of Python 3.6, it will stop security support in 7 months, which I don't think it's a near future. In order to leverage the new `DataLoader` with `DataPipe` to all users no matter which Python version they are using, I have to make sure this DataPipe typing system has the compatibility for Python 3.6.
